### PR TITLE
remove /var/lib/docker volume from containerized node

### PR DESCRIPTION
**Pull request by sjenning**
xref https://bugzilla.redhat.com/show_bug.cgi?id=1467824

tl;dr upstream docker does not allow the forced removal of containers anymore.  Before, a forced removal, when using devicemapper graph driver, simply orphaned a thin device and those devices would eventually consume the pool.  The upstream docker change was backported to docker-1.12.6-40 to fix this issue.

However, the mounting of `/var/lib/docker` into the containerized node is causing other openshift services from being able to (re)start successfully since the systemd unit files first remove the old container, which fails due to the containerized node holding the mount point, then tries a docker run, which fails because a container with that name already exists.

To avoid this, the PR removes `/var/lib/docker` from the volumes mounted into the containerized node.

@derekwaynecarr @smarterclayton @eparis

### CODE REVIEW

**Comment by sjenning** [comment]
@rhvgoyal just saw your update on the BZ.  Is this not needed anymore?

**Comment by rhvgoyal** [comment]
@sjenning I have created a PR so that oci-umount plugin will take care of removing multiple mounts inside container. But I think modifying unit file to not mount /var/lib/docker/ to begin with is a good idea. 

IOW, while my PR to oci-umount will take care of this, but we should merge this change anyway (irrespective of my PR).

**Comment by rhvgoyal** [comment]
https://github.com/projectatomic/oci-umount/pull/10

**Comment by sjenning** [comment]
[test]

**Comment by smarterclayton** [comment]
@csrwng do we need to change cluster up?

**Comment by csrwng** [comment]
@smarterclayton yes we need to remove it there as well, will create a PR

**Comment by csrwng** [comment]
@sjenning @rhvgoyal under what circumstances should we not mount /var/lib/docker anymore? Is it based on the Docker version? or the version of the kubelet? Will things break, for example, if I try to run origin v1.5.1 images on top of Docker docker-1.12.6-30 and I don't mount /var/lib/docker ?

**Comment by rhvgoyal** [comment]
@csrwng I am expecting that you don't have to mount /var/lib/docker/ in any circumstance at all. There is one logging case of fluentd which still needs it. Apart from that, if there are other use cases, I want to know.

**Comment by csrwng** [comment]
@rhvgoyal So for the one fluentd case, is it no longer required?

**Comment by rhvgoyal** [comment]
@rhvgoyal I think for fluentd case you still require it. As I understand it, for json log based driver, they still go trough container logs under /var/lib/docker.

**Comment by rhvgoyal** [comment]
@csrwng I myself am trying to figure out who else needs it and what breaks if we don't mount it.

**Comment by giuseppe** [comment]
would a build still work?  I've tried this change in the system container and if there is no `/var/lib/docker/containers` present, the build fails with:

```
# oc logs -f bc/hello-openshift-plus
failed to open log file "/var/log/pods/7267b90a-65dc-11e7-9b7c-525400874d80/docker-build_0.log": open /var/log/pods/7267b90a-65dc-11e7-9b7c-525400874d80/docker-build_0.log: no such file or directory[root@69ccc7cff516 /]# oc logs -f bc/hello-openshift-plus
```

That happens since files under `/var/log/pods` are symbolic links to `/var/lib/docker/containers/`, e.g.

```
ls -l /var/log/pods/bd5e4ab0-65df-11e7-80e7-525400874d80/docker-build_0.log
lrwxrwxrwx. 1 root root 165 Jul 11 02:22 /var/log/pods/bd5e4ab0-65df-11e7-80e7-525400874d80/docker-build_0.log -> /var/lib/docker/containers/37eb94514154a15d542856c03a66cd08fc0d010a1ff9b327e79ceffffe8b4a29/37eb94514154a15d542856c03a66cd08fc0d010a1ff9b327e79ceffffe8b4a29-json.log
```

I have not tested this with the Docker container, so please ignore my comment if it doesn't happen there.  If so though, what is done differently than files in `/var/log/pods` have the right content?

**Comment by rhvgoyal** [comment]
@giuseppe These seems to be just accessing json logs of a container. /var/lib/docker/containers/ is just a bind mount onto itself. So even if I unmount it, logs are still there.

So I doubt this breakage is due to unmounting of /var/lib/docker/containers. That directory and logs with-in should still be accessible.

**Comment by giuseppe** [comment]
@rhvgoyal yes true, my point was that if we just drop the `/var/lib/docker` mount point (as currently the patch seems to do) we lose completely access to `/var/lib/docker/containers` from the node container.  

Do we need to add a mount point for `/var/lib/docker/containers` since `/var/lib/docker` is not mounted anymore or how that will be accessible?  I needed this in the system container case to fix the issue I reported above.

**Comment by rhvgoyal** [comment]
I don't understand openshift. So you will have to simplify it for me. So why do we need access to logs while doing a build?

**Comment by giuseppe** [comment]
I am not expert either but my guess is that it is needed for `oc logs`.  If I use `oc logs` I can see the docker build log there.

**Comment by rhvgoyal** [comment]
Ok, so this is the case where "oc logs" is being run from inside a container and it is trying to access logs of a container? This sounds similar to fluentd accessing container logs.

So why are we not using "docker logs" to access logs?

**Comment by rhvgoyal** [comment]
Anyway, if some container does need /var/lib/docker, then yes they will have to volume mount it in. I don't know which containers need it. So openshift service/container needs to decide which containers really need it. 

oci-umount should take care of unmounting /var/lib/docker/devicemapper and /var/lib/docker/containers from inside the containers.

**Comment by smarterclayton** [comment]
@derekwaynecarr @sjennings can you help Vivek at all with questions about
what the kubelet needs access to in order to service logs?

**Comment by sjenning** [comment]
@smarterclayton I was trying to prove by doing but containerized install is busted https://github.com/openshift/origin/issues/15147.  I'll have to look at the code in the meantime.

**Comment by sjenning** [comment]
This is the only thing I can find

https://github.com/openshift/origin/blob/master/pkg/cmd/server/kubernetes/node/node.go#L61-L63

For that, `/var/lib/docker` doesn't have to be mounted in, but it does need to exist.

**Comment by sjenning** [comment]
@rhvgoyal do we still need this?

**Comment by sjenning** [comment]
I think this is handled via other means now
